component/kafka/simple/consumer.go

```diff
@@ -0,0 +1,301 @@
+	for {
+		select {
+		case msg, ok := <-ch:
+			if !ok {
+				log.Debug("messages channel closed")
+				return nil
+			}
+			log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+			messageStatusCountInc(messageReceived, partition, msg.Topic)
+			err := c.unit(ctx, msg)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if errors.Is(ctx.Err(), context.Canceled) {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) consumePartitionBatch(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) (err error) {
+	// We need to enforce flushing the buffer before to return (e.g., context cancelled but messages are still buffered)
+	defer func() {
+		errFlush := c.flush(ctx)
+		if errFlush != nil {
+			err = errFlush
+		}
+	}()
+
+	for {
+		select {
+		case msg, ok := <-ch:
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+				messageStatusCountInc(messageReceived, partition, msg.Topic)
+				err := c.insertMessage(ctx, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
+				return nil
+			}
+		case <-c.ticker.C:
+			err := c.flush(ctx)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if errors.Is(ctx.Err(), context.Canceled) {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) unit(ctx context.Context, msg *sarama.ConsumerMessage) error {
+	messageStatusCountInc(messageProcessed, msg.Partition, msg.Topic)
+	ctx, sp := c.getContextWithCorrelation(ctx, msg)
+	messages := []kafka.Message{kafka.NewMessage(ctx, sp, msg)}
+
+	btc := kafka.NewBatch(messages)
+	err := c.component.proc(btc)
+	if err != nil {
+		if errors.Is(ctx.Err(), context.Canceled) {
+			return fmt.Errorf("context was cancelled after processing error: %w", err)
+		}
+		err = c.executeFailureStrategy(messages, err)
+		if err != nil {
+			return err
+		}
+	} else {
+		c.mu.Lock()
+		c.hasProcessedMessages = true
+		c.updateLatestOffsetReached(messages)
+		c.mu.Unlock()
+	}
+
+	trace.SpanSuccess(messages[0].Span())
+
+	return nil
+}
+
+func (c *consumerHandler) insertMessage(ctx context.Context, msg *sarama.ConsumerMessage) error {
+	c.mu.Lock()
+	c.msgBuf = append(c.msgBuf, msg)
+	if len(c.msgBuf) >= int(c.component.batchSize) {
+		c.mu.Unlock()
+		return c.flush(ctx)
+	}
+	c.mu.Unlock()
+	return nil
+}
+
+func (c *consumerHandler) flush(ctx context.Context) error {
+	c.mu.Lock()
+	if len(c.msgBuf) == 0 {
+		c.mu.Unlock()
+		return nil
+	}
+	messages := make([]kafka.Message, len(c.msgBuf))
+	for i, msg := range c.msgBuf {
+		messageStatusCountInc(messageProcessed, msg.Partition, msg.Topic)
```

Comment:
Similar comment to the one in `unit()` At this point we are preparing the messages but they won't be processed until the batch starts being processed.
Based on the name of the event, I'd expect this to happen close to `c.hasPrecessedMessages=true` assignment